File: 2024/All/Arrays/Feb/22Feb/3sum.js
let arr = [-1, 0, 1, 2, -1, -4];

let n = arr.length;

const threeSum = (arr, n) => {
  let ans = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      for (let k = j + 1; k < n; k++) {
        if (arr[i] + arr[j] + arr[k] === 0) {
          let temp = [arr[i], arr[j], arr[k]];
          temp.sort((a, b) => a - b);
          ans.push(temp);
        }
      }
    }
  }
  let set = new Set(ans.map(JSON.stringify));
  ans = Array.from(set).map(JSON.parse);

  return ans
 };

// console.log(threeSum(arr, n));

const result = threeSum(arr,n)

for (let i = 0; i < result.length; i++) {
     
  }

// console.log(result)

/// Better Approach

let ans = [];
arr.sort((a, b) => a - b);
for (let i = 0; i < n; i++) {
    // remove duplicates:
    if (i !== 0 && arr[i] === arr[i - 1]) continue;

    // moving 2 pointers:
    let j = i + 1;
    let k = n - 1;
    while (j < k) {
        let sum = arr[i] + arr[j] + arr[k];
        if (sum < 0) {
            j++;
        } else if (sum > 0) {
            k--;
        } else {
            let temp = [arr[i], arr[j], arr[k]];
            ans.push(temp);
            j++;
            k--;
            // skip the duplicates:
            while (j < k && arr[j] === arr[j - 1]) j++;
            while (j < k && arr[k] === arr[k + 1]) k--;
        }
    }
}
return ans;

File: 2024/Cpp/Arrays/3sum.cpp
// Brute force 
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
 
         set <vector<int>> st;

        for(int i = 0; i < n; i++){
            for(int j = i+1; j<n; j++){
                for(int k = j+1 ; k<n; k++){
                    if(nums[i]+nums[j]+nums[k] == 0){
                        vector<int>temp = {nums[i],nums[j],nums[k]};
                        sort(temp.begin(),temp.end());
                        st.insert(temp);
                    }
                }
            }
        }
        vector<vector<int>> ans(st.begin(),st.end());
        return ans;
    }
};

//Better - n^2
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
 
        set<vector<int>> st;
        for(int i = 0; i < n - 2; i++) {
            set<int> hashSet;  
            for(int j = i+1; j < n; j++) {
                int third = -(nums[i] + nums[j]);
                if (hashSet.find(third) != hashSet.end()) {
                    vector<int> temp = {nums[i], nums[j], third};
                    sort(temp.begin(), temp.end());
                    st.insert(temp);
                }
                hashSet.insert(nums[j]);
            }
        }
        vector<vector<int>> ans(st.begin(), st.end());
        return ans;
    }
};


// using 2pointers 

